Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,628 @@
+                           name: .ProfileDidFinishSyncing,
+                           object: nil)
+    }
+
+    func doInBackgroundAfter(_ millis: Int64, _ block: @escaping () -> Void) {
+        let queue = DispatchQueue.global(qos: DispatchQoS.background.qosClass)
+        // Pretty ambiguous here. I'm thinking .now was DispatchTime.now() and not
+        // Date.now()
+        queue.asyncAfter(
+            deadline: DispatchTime.now() + DispatchTimeInterval.milliseconds(Int(millis)),
+            execute: block)
+    }
+
+    @objc func onStartSyncing(_ notification: NSNotification) {
+        syncDisplayState = .inProgress
+    }
+
+    @objc func onFinishSyncing(_ notification: NSNotification) {
+        if let syncState = syncDisplayState, syncState == .good {
+            self.lastSyncFinishTime = Date.now()
+        }
+    }
+
+    var prefsForSync: Prefs {
+        return self.prefs.branch("sync")
+    }
+
+    public func onAddedAccount() -> Success {
+        // Only sync if we're green lit. This makes sure that we don't sync unverified
+        // accounts.
+        guard self.profile.hasSyncableAccount() else { return succeed() }
+
+        self.beginTimedSyncs()
+        return self.syncEverything(why: .didLogin)
+    }
+
+    public func onRemovedAccount() -> Success {
+        let clearPrefs: () -> Success = {
+            withExtendedLifetime(self) {
+                // Clear prefs after we're done clearing everything else -- just in case
+                // one of them needs the prefs and we race. Clear regardless of success
+                // or failure.
+
+                // This will remove keys from the Keychain if they exist, as well
+                // as wiping the Sync prefs.
+
+                // XXX: `Scratchpad.clearFromPrefs` and `clearAll` were pulled from
+                // `SyncStateMachine.clearStateFromPrefs` to reduce RustSyncManager's
+                // dependence on the swift sync state machine logid. This will make
+                // refactoring or eliminating that code easier once the rust sync manager
+                // rollout is complete.
+                Scratchpad.clearFromPrefs(self.prefsForSync.branch("scratchpad"))
+                self.prefsForSync.clearAll()
+            }
+            return succeed()
+        }
+        syncQueue.async { [unowned self] in
+            self.syncManagerAPI.disconnect()
+        }
+        return clearPrefs()
+    }
+
+    private func getEngineEnablementChangesForAccount() -> [String: Bool] {
+        var engineEnablements: [String: Bool] = [:]
+        // We just created the account, the user went through the Choose What to Sync
+        // screen on FxA.
+        if let declined = UserDefaults.standard.stringArray(
+            forKey: "fxa.cwts.declinedSyncEngines") {
+            declined.forEach { engineEnablements[$0] = false }
+            UserDefaults.standard.removeObject(forKey: "fxa.cwts.declinedSyncEngines")
```

Comment:
nit, Looks like this key is used in other places, we could maybe have it as a constant?